README.md
JoliSymfony - Symfony2
=====================
[![NPM Version](https://img.shields.io/npm/v/generator-joli-symfony.svg?style=flat-square)](https://www.npmjs.com/package/generator-joli-symfony)
[![Download Month](https://img.shields.io/npm/dm/generator-joli-symfony.svg?style=flat-square)](https://www.npmjs.com/package/generator-joli-symfony)

generator-joli-symfony is a [Yeoman Generator](http://yeoman.io/generators/) to scaffold Symfony2 projects with sensible defaults, common bundles and frontend tools.

It will create a new Symfony project, remove Assetic and replace it with either Gulp, Grunt or Brunch.

### Demo

![](http://i.imgur.com/idCv3zq.gif)

## Dependencies

Mandatory dependencies :

- [composer](https://getcomposer.org/download/)
- [npm](http://nodejs.org/)
- [yo](http://yeoman.io/)
- [gulp](https://github.com/gulpjs/gulp/blob/master/docs/getting-started.md#getting-started)
- [grunt](http://gruntjs.com/getting-started)
- [bower](http://bower.io/#install-bower)
- [brunch](http://brunch.io/)

Optional dependencies :

- [gem](https://www.ruby-lang.org/en/downloads/)
- [compass](http://compass-style.org/install/)


## What you can choose

* Symfony 2 Standard Edition:
 * [2.3](https://github.com/symfony/symfony-standard/tree/2.3)
 * [2.6](https://github.com/symfony/symfony-standard/tree/2.6)
 * [2.7](https://github.com/symfony/symfony-standard/tree/2.7)
* Grunt:
 * [grunt-compass](https://github.com/gruntjs/grunt-contrib-compass)
 * [grunt-less](https://github.com/gruntjs/grunt-contrib-less)
 * [grunt-babel](https://github.com/babel/grunt-babel)
 * [grunt-coffee](https://github.com/gruntjs/grunt-contrib-coffee)
 * [grunt-typescript](https://github.com/k-maru/grunt-typescript)
* Gulp:
 * [gulp-ruby-sass](https://github.com/sindresorhus/gulp-ruby-sass)
 * gulp-copy
 * gulp-javascript
 * [gulp-less](https://github.com/plus3network/gulp-less)
 * [gulp-babel](https://github.com/babel/gulp-babel)
 * [gulp-typescript](https://github.com/ivogabe/gulp-typescript)
 * [gulp-coffee](https://github.com/wearefractal/gulp-coffee)
* Brunch:
 * [less-brunch](https://github.com/brunch/less-brunch)
 * [sass-brunch](https://github.com/brunch/sass-brunch)
 * [stylus-brunch](https://github.com/brunch/stylus-brunch)
 * [coffee-script-brunch](https://github.com/brunch/coffee-script-brunch)
 * [typescript-brunch](https://github.com/joshheyse/typescript-brunch)
 * [uglify-js-brunch](https://github.com/brunch/uglify-js-brunch)
 * [babel-brunch](https://github.com/babel/babel-brunch)
* Symfony2 Bundle:
 * [DoctrineFixturesBundle](https://github.com/doctrine/DoctrineFixturesBundle)
 * [DoctrineMigrationsBundle](https://github.com/doctrine/DoctrineMigrationsBundle)
* Bootstrap-sass-official

## Default workflow

* Installs Symfony
* Removes Assetic
 * Only if using grunt-less or grunt-babel or grunt-compass: **(grunt-contrib-watch)**
 * Only if using grunt-typescript or grunt-coffeee or grunt-babel : **(grunt-uglify)**
 * Only if using gulp-ruby-sass **(gulp-watch)**
 * Only if using gulp-ruby-sass or gulp-javascript **(gulp-rename, gulp-sourcemaps)**
 * Only if using gulp-ruby-sass or gulp-less **(gulp-minify-css)**
 * Only if using gulp-javascript or gulp-coffee or gulp-typescript or gulp-babel **(gulp-uglify, gulp-concat)**
* Starts the automatic execution of `bower`, `npm` and `composer` after scaffolding has finished.

## Assets location

Assets are stored in the **app/Resources/** folder :

* app/Resources/scss
* app/Resources/fonts
* app/Resources/...

## Getting Started

- Install: `npm install -g yo`
- Install: `npm install -g generator-joli-symfony`
- Run: `yo joli-symfony --force`

## Contribute

`generator-joli-symfony` is fork-friendly : you can maintain a custom version, in which you `npm install && npm link` so that you can still use it with `yo joli-symfony`, or with the name of your choice.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2015-08-06

### Changed
- Bootstrap's installation improved
- Assetic's deletion improved
- Brand new method to fetch Symfony's versions
- Now detects if `composer` and `bower`are already installed on the system
- `Gruntfile.js` and `Gulpfile.js` updated
- `bower.json` updated
- Npm modules updated

### Removed
- Module `grunt-sass` removed
- Module `load-grunt-tasks` removed
- Module `gulp-load-plugins` removed
- Bundle `DoctrineMongoDBBundle` removed
- File `.bowerrc` removed - All files installed with bower will be stored in the default `bower_component` directory.


### Added

Grunt
- Module `grunt-less` added
- Module `grunt-babel` added
- Module `grunt-contrib-watch` added

Gulp
- Module `gulp-less` added
- Module `gulp-babel` added
- Module `gulp-typescript` added
- Module `gulp-coffee` added
- Module `gulp-watch` added

Brunch
- Configuration file `brunch-config.js` added
- Choice `Brunch` added in the tools list
- Module `less-brunch` added
- Module `sass-brunch` added
- Module `stylus-brunch` added
- Module `coffee-brunch` added
- Module `typescript-brunch` added
- Module `uglify-js-brunch` added
- Module `babel-brunch` added

NPM
- Module `fs-extra` added

- Demo added
	- files `scss`
	- files `less`
	- files `stylus`
	- files `Typescript`
	- files `Coffee`
	- files `Js (export)`
	- files `Js (Class ES6)`


test/test-app.js
'use strict';

var path = require('path');
var assert = require('yeoman-generator').assert;
var helpers = require('yeoman-generator').test;
var os = require('os');

describe('jolistarter:app', function () {
  before(function (done) {
    helpers.run(path.join(__dirname, '../app'))
      .inDir(path.join(os.tmpdir(), './temp-test'))
      .withOptions({ 'skip-install': true })
      .withPrompt({
        someOption: true
      })
      .on('end', done);
  });

  it('creates files', function () {
    assert.file([
      'bower.json',
      'package.json',
      '.editorconfig',
      '.jshintrc'
    ]);
  });
});


app/templates/_Gulpfile.js
var gulp = require('gulp');
<% if (gulpRubySass) { %>var sass = require('gulp-ruby-sass');<% } %>
<% if (gulpRubySass || gulpLess) { %>
var sourcemaps = require('gulp-sourcemaps');
var minifyCss = require('gulp-minify-css');
var watch = require('gulp-watch');<% } %>
<% if (gulpRubySass || gulpConcat || gulpLess) { %>
var rename = require('gulp-rename');<% } %>
<% if (gulpConcat || gulpBabel || gulpCoffee || gulpTypescript) { %>
var concat = require('gulp-concat');
var uglify = require('gulp-uglify');<% } %>
<% if (gulpTypescript) { %>var ts = require('gulp-typescript');<% } %>
<% if (gulpCoffee) { %>var coffee = require('gulp-coffee');<% } %>
<% if (gulpLess) { %>var less = require('gulp-less');<% } %>
<% if (gulpBabel) { %>var babel = require('gulp-babel');<% } %>
var gutil = require('gulp-util');

var path = {
  app: 'app/Resources',
  bower_components: './bower_components'
};

<% if (gulpCopy) { %>
gulp.task('copy', function() {
    gulp.src(path.app + '/assets/fonts/*.{ttf,woff,eof,svg,eot}')
        .pipe(gulp.dest('web/fonts/'));
});<% } %>

<% if (gulpRubySass) { %>
/**
* gulp-ruby-sass
* @see https://www.npmjs.com/package/gulp-ruby-sass
*
* Compile Sass to CSS using Compass.
*/
gulp.task('sass', function() {

  return sass(path.app + '/scss', { compass: true, style: 'compressed', sourcemap: true })
    .on('error', function (err) {
      console.error('Error!', err.message);
    })
    .pipe(minifyCss({keepSpecialComments:0}))
    .pipe(sourcemaps.write())
    .pipe(rename({suffix: '.min'}))
    .pipe(gulp.dest('web/css/'));
});<% } %>

<% if (gulpConcat) { %>
/**
* gulp-concat && gulp-uglify
* @see https://www.npmjs.com/package/gulp-concat
* @see https://www.npmjs.com/package/gulp-uglify
*
* Compile and minify js vendor (bower_components).
*/
gulp.task('vendor', function() {
    gulp.src([
        './bower_components/jquery/dist/jquery.js',
        './bower_components/bootstrap/assets/javascripts/bootstrap.js'
    ])
    .pipe(concat('vendor.js'))
    .pipe(uglify({mangle: true}))
    .pipe(rename({suffix: '.min'}))
    .pipe(gulp.dest('web/js/'));
});

/**
* gulp-concat && gulp-uglify
* @see https://www.npmjs.com/package/gulp-concat
* @see https://www.npmjs.com/package/gulp-uglify
*
* Compile and minify js App (app/Resources/js).
*/
gulp.task('app', function() {
    return gulp.src(path.app + '/js/**/*.js')
    .pipe(sourcemaps.init())
    .pipe(concat('app.js'))
    .pipe(uglify({mangle: true}).on('error', gutil.log))
    .pipe(rename({suffix: '.min'}))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('web/js/'));
});<% } %>

<% if (gulpLess) { %>
/**
* gulp-less
* @see https://www.npmjs.com/package/gulp-less
*
* Compile Less to CSS.
*/
gulp.task('less', function() {
    gulp.src(path.app + '/less/main.less')
        .pipe(sourcemaps.init())
        .pipe(less())
        .pipe(minifyCss({keepSpecialComments:0}))
        .pipe(sourcemaps.write())
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('web/css/'));
});<% } %>

<% if (gulpCoffee) { %>
/**
* gulp-coffee
* @see https://www.npmjs.com/package/gulp-coffee
*
* Compile CoffeeScript files to Javascript.
*/
gulp.task('coffee', function() {
  gulp.src(path.app + '/js/*.coffee')
    .pipe(sourcemaps.init())
    .pipe(coffee({ bare: true })).on('error', gutil.log)
    .pipe(concat('coffee.js'))
    .pipe(uglify({mangle: true}).on('error', gutil.log))
    .pipe(rename({suffix: '.min'}))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('web/js/'));
});<% } %>

<% if (gulpTypescript) { %>
/**
* gulp-typescript
* @see https://www.npmjs.com/package/gulp-typescript
*
* Compile Typescript files to Javascript.
*/
gulp.task('ts', function() {
  var tsResult = gulp.src(path.app + '/js/*.ts')
    .pipe(sourcemaps.init())
    .pipe(ts({
      sortOutput: true,
      target: 'ES6', // 'ES3' (default), 'ES5' or 'ES6'.
      module: 'commonjs' //'commonjs' or 'amd'.
    }));

  return tsResult.js
    .pipe(concat('typescript.js'))
    .pipe(uglify({mangle: true}).on('error', gutil.log))
    .pipe(rename({suffix: '.min'}))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('web/js/'));
});<% } %>

<% if (gulpBabel) { %>
/**
* gulp-babel
* @see https://www.npmjs.com/package/gulp-babel
*
* Turn ES6 code into vanilla ES5 with no runtime required
*/
gulp.task('babel', function () {
  return gulp.src(path.app + '/js/**/*.js')
    .pipe(sourcemaps.init())
    .pipe(babel())
    .pipe(concat('babel.js'))
    .pipe(uglify({mangle: true}).on('error', gutil.log))
    .pipe(rename({suffix: '.min'}))
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest('web/js/'));
});<% } %>

gulp.task('default', [
<% if (gulpCopy) { %>'copy',<% } %>
<% if (gulpConcat) { %>'vendor',<% } %>
<% if (gulpConcat) { %>'app',<% } %>
<% if (gulpBabel) { %>'babel',<% } %>
<% if (gulpCoffee) { %>'coffee',<% } %>
<% if (gulpTypescript) { %>'ts',<% } %>
<% if (gulpLess) { %>'less',<% } %>
<% if (gulpRubySass) { %>'sass'<% } %>
]);


app/templates/_brunch-config.js
'use strict';

exports.config = {
  paths: {
    'public': 'web',
    'watched': ['app/Resources']
  },
  files: {
    javascripts: {
      joinTo: {
        'js/app.js': /^app/,
        'js/vendor.js': /^(?!app)/,
      }
    },
    stylesheets: {
      joinTo: 'css/style.css'
    }
  },
  conventions: {
    ignored: [
      /\/_/, // File begining by "_" like _settings.scss
      // Brunch does include all Bower components by default, we blacklist unneeded ones.
      //'bower_components/foundation/'
    ],
    assets: /^app\/Resources\/assets/
  },
  plugins: {
    <% if (babelBrunch) { %>
    babel: {
      pattern: /\.(js|jsx)$/
    },<% } %><% if (sassBrunch) { %>
    sass: {
      allowCache: true
    },<% } %><% if (uglifyJsBrunch) { %>
    uglify: {
      mangle: true,
      compress: {
        global_defs: {
          DEBUG: false
        }
      }
    },<% } %><% if (typescriptBrunch) { %>
    Typescript: {
      tscArgs: '--module commonjs'
    },<% } %>
    cleancss: {
      keepSpecialComments: 0,
      removeEmpty: true
    }
  }
};

app/index.js
"use strict";

var yeoman = require('yeoman-generator');
var chalk = require('chalk');
var path = require('path');
var yaml = require('js-yaml');
var fs = require('fs-extra');
var rmdir = require('rimraf');
var child_process = require('child_process');
var http = require("http");

module.exports = yeoman.generators.Base.extend({
  initializing: function () {
    this.pkg = require('../package.json');
    var Jolicode = '\n        ' + chalk.yellow('##') + '       ##            ##  ##   #######                   ##                ' + chalk.yellow('##') + '  ' + chalk.yellow('##') + '  \n      ' + chalk.yellow('##') + '         ##            ##      ##     ##                  ##               ' + chalk.yellow('##') + '    ' + chalk.yellow('##') + '  \n     ' + chalk.yellow('##') + '          ##   ######   ##  ##  ##          ######    #######   ######     ' + chalk.yellow('##') + '      ' + chalk.yellow('##') + '  \n     ' + chalk.yellow('##') + '          ##  ##    ##  ##  ##  ##         ##    ##  ##    ##  ########   ' + chalk.yellow('##') + '       ' + chalk.yellow('##') + '  \n      ' + chalk.yellow('##') + '  ##    ###  ##    ##  ##  ##  ##     ##  ##    ##  ##    ##  ##        ' + chalk.yellow('##') + '       ' + chalk.yellow('##') + '  \n       ' + chalk.yellow('##') + '   #####     ######   ##  ##   #######    ######    #######   ######  ' + chalk.yellow('##') + '       ' + chalk.yellow('##');
    var JolicodeDesc = '\n\n   A Yeoman generator for the Symfony2 framework\n\n   Created by ' + chalk.yellow('@JoliCode ') + ' & ' + chalk.blue('@lbrunet_com') + '\n   ' + chalk.cyan('http://jolicode.com/') + '\n';
    this.log(Jolicode);
    this.log(JolicodeDesc);
  },

  askSymfonyStandard: function () {
    var done = this.async();

    this.SymfonyStandardDistribution = {
      host: 'http://symfony.com/download?v=Symfony_Standard_Vendors_',
      commit: '2.7',
      ext: 'zip'
    };

    var prompts = [{
      type: 'confirm',
      name: 'symfonyStandard',
      message: 'Would you like to use the Symfony "Standard Edition" distribution ' + this.SymfonyStandardDistribution.commit,
      default: true
    }];

    this.prompt(prompts, function(answers) {

      if (answers.symfonyStandard) {
        this.symfonyDistribution = this.SymfonyStandardDistribution;
      } else {
        this.symfonyDistribution = null;
      }

      done();
    }.bind(this));
  },

  askSymfonyCustom: function () {
    if (this.symfonyDistribution === null) {
      var done = this.async();

      console.log('Please provide GitHub details of the Symfony distribution you would like to use.');
      console.log('e.g. http://github.com/symfony/symfony-standard/tree/[commit].');

      var prompts = [{
        type: 'list',
        name: 'symfonyCommit',
        message: 'Commit (commit/branch/tag)',
        default: '2.7',
        choices: ['2.3', '2.6', '2.7']
      }];

      this.prompt(prompts, function(answers) {

        var repo = 'https://github.com/' + 'symfony' + '/' + 'symfony-standard' + '/tree/' + answers.symfonyCommit;
        console.log('Thanks! I\'ll use ' + repo);
        console.log('');

        this.symfonyDistribution = {
          host: 'http://symfony.com/download?v=Symfony_Standard_Vendors_',
          commit: answers.symfonyCommit,
          ext: 'zip'
        };

        done();
      }.bind(this));
    }
  },

  getTagSymfony: function() {
    var version = this.symfonyDistribution.commit,
        done = this.async();

    http.get('http://symfony.com/versions.json', function(res) {

     var data = '';

     res.on('data', function(d) {
       data += d;
     });

     res.on('error', function(e) {
       console.log("Got error: " + e.message);
     });

     res.on('end', function() {
       var parsed = JSON.parse(data);
       this.symfonyDistribution.commit = parsed[version];
       done();
     }.bind(this));
    }.bind(this));
  },

  askToolsExtension: function() {
    var done = this.async();
    var prompts = [{
        type: 'list',
        name: 'toolsExtension',
        message: 'Which tools would you like to use?',
        default: 'gulp',
        choices: ['grunt', 'gulp', 'brunch']
    }];
    this.prompt(prompts, function(answers) {
      this.toolsExtension = answers.toolsExtension;
      done();
    }.bind(this));
  },

  askGruntCustom: function () {
    if (this.toolsExtension === 'grunt') {
      var done = this.async();

      var prompts = [{
        type: 'checkbox',
        name: 'gruntCustom',
        message: 'Customize Gruntfile',
        choices: [
          {
            name: 'grunt-compass',
            value: 'gruntCompass',
            checked: true
          },
          {
            name: 'grunt-less',
            value: 'gruntLess',
            checked: false
          },
          {
            name: 'grunt-babel' + chalk.yellow(' => Turn ES6 code into vanilla ES5 with no runtime required'),
            value: 'gruntBabel',
            checked: true
          },
          {
            name: 'grunt-coffee',
            value: 'gruntCoffee',
            checked: false
          },
          {
            name: 'grunt-typescript',
            value: 'gruntTypescript',
            checked: true
          }
        ]
      }];

      this.prompt(prompts, function(answers) {
        function hasFeature(feat) {
          return answers.gruntCustom.indexOf(feat) !== -1;
        }

        this.gruntcompass = hasFeature('gruntCompass');
        this.gruntcoffee = hasFeature('gruntCoffee');
        this.gruntTypescript = hasFeature('gruntTypescript');
        this.gruntLess = hasFeature('gruntLess');
        this.gruntBabel = hasFeature('gruntBabel');

        done();
      }.bind(this));
    }
  },

  askGulpCustom: function() {
    if (this.toolsExtension === 'gulp') {
      var done = this.async();

      var prompts = [{
        type: 'checkbox',
        name: 'gulpCustom',
        message: 'Customize Gulpfile',
        choices: [
          {
            name: 'gulp-ruby-sass '  + chalk.yellow(' ♥ '),
            value: 'gulpRubySass',
            checked: true
          },
          {
            name: 'gulp-copy' + chalk.red(' => Do not use the gulp-copy module but a task default !'),
            value: 'gulpCopy',
            checked: false
          },
          {
            name: 'gulp-javascript' + chalk.yellow(' => use for javascript ES5, ES4, ES3, not ES6 !'),
            value: 'gulpConcat',
            checked: false
          },
          {
            name: 'gulp-less',
            value: 'gulpLess',
            checked: false
          },
          {
            name: 'gulp-babel ' + chalk.yellow(' => Turn ES6 code into vanilla ES5 with no runtime required ♥'),
            value: 'gulpBabel',
            checked: true
          },
          {
            name: 'gulp-typescript',
            value: 'gulpTypescript',
            checked: false
          },
          {
            name: 'gulp-coffee',
            value: 'gulpCoffee',
            checked: false
          }
        ]
      }];

      this.prompt(prompts, function (answers) {
        function hasFeature(feat) {
          return answers.gulpCustom.indexOf(feat) !== -1;
        }

        this.gulpRubySass = hasFeature('gulpRubySass');
        this.gulpCopy = hasFeature('gulpCopy');
        this.gulpConcat = hasFeature('gulpConcat');
        this.gulpLess = hasFeature('gulpLess');
        this.gulpBabel = hasFeature('gulpBabel');
        this.gulpTypescript = hasFeature('gulpTypescript');
        this.gulpCoffee = hasFeature('gulpCoffee');

        done();
      }.bind(this));
    }
  },

  askBrunchCustom: function() {
    if (this.toolsExtension === 'brunch') {
      var done = this.async();

      var prompts = [{
        type: 'checkbox',
        name: 'brunchCustom',
        message: 'Customize Brunch',
        choices: [
          {
            name: 'less-brunch',
            value: 'lessBrunch',
            checked: false
          },
          {
            name:  'sass-brunch ' + chalk.yellow(' ♥ '),
            value: 'sassBrunch',
            checked: true
          },
          {
            name: 'stylus-brunch',
            value: 'stylusBrunch',
            checked: false
          },
          {
            name: 'coffee-script-brunch',
            value: 'coffeeScriptBrunch',
            checked: false
          },
          {
            name: 'typescript-brunch',
            value: 'typescriptBrunch',
            checked: false
          },
          {
            name: 'uglify-js-brunch',
            value: 'uglifyJsBrunch',
            checked: true
          },
          {
            name: 'babel-brunch' + chalk.yellow(' => Turn ES6 code into vanilla ES5 with no runtime required ♥'),
            value: 'babelBrunch',
            checked: true
          }
        ]
      }];

      this.prompt(prompts, function (answers) {
        function hasFeature(feat) {
          return answers.brunchCustom.indexOf(feat) !== -1;
        }

        this.lessBrunch = hasFeature('lessBrunch');
        this.sassBrunch = hasFeature('sassBrunch');
        this.stylusBrunch = hasFeature('stylusBrunch');
        this.coffeeScriptBrunch = hasFeature('coffeeScriptBrunch');
        this.typescriptBrunch = hasFeature('typescriptBrunch');
        this.uglifyJsBrunch = hasFeature('uglifyJsBrunch');
        this.babelBrunch = hasFeature('babelBrunch');

        done();
      }.bind(this));
    }
  },

  askBootStrapSass: function() {
    var done = this.async();

    var prompts = [{
      type: 'confirm',
      name: 'bootStrapSass',
      message: 'Would you like to use "BootStrap Sass"?',
      default: true
    }];

    this.prompt(prompts, function (answers) {
      this.bootStrapSass = answers.bootStrapSass;
      done();
    }.bind(this));
  },

  askbundle: function() {
    var done = this.async();

    var prompts = [{
      type: 'checkbox',
      name: 'addBundle',
      message: 'Which bundle would you like to use?',
      choices: [
      {
        name: 'DoctrineFixturesBundle',
        value: 'fixturebundle',
        checked: true
      },
      {
        name: 'DoctrineMigrationsBundle',
        value: 'migrationbundle',
        checked: false
      }
      ]

    }];

    this.prompt(prompts, function(answers) {
      function hasFeature(feat){
        return answers.addBundle.indexOf(feat) !== -1;
      }

      this.fixturebundle = hasFeature('fixturebundle');
      this.migrationbundle = hasFeature('migrationbundle');
      done();
    }.bind(this));
  },

    symfonyBase: function() {
    var done = this.async();
    var appPath = this.destinationRoot();
    var repo = this.symfonyDistribution.host + this.symfonyDistribution.commit  + '.' + this.symfonyDistribution.ext;

    this.extract(repo, appPath, function (err, remote) {
          if (err) {
            return done(err);
          } else {
            console.log(' 👍 ' + chalk.green(' Download success ! '));
            done();
          }
        }
    );
  },

  moveSymfonyBase: function() {
    var done = this.async();
    var directory = this.destinationRoot() + '/Symfony';
    this.directory(directory, '.');
    fs.move('./Symfony/', '.', function (err) {
      if (err) return console.error(err)
      console.log("success!");
    });
    done();
  },

  removeSymfonyBase: function() {
    var done = this.async();
    var directory = this.destinationRoot() + '/Symfony';
    rmdir(directory, function(error) {
      if (null === error) {
        console.log(' 👍 ' + chalk.yellow(' Installation Symfony success !'));
      }
    });
    done();
  },

  writing: {
    app: function () {
      if (this.toolsExtension === 'grunt') {
        this.template('_Gruntfile.js', 'Gruntfile.js');
      }
      if (this.toolsExtension === 'gulp') {
        this.template('_Gulpfile.js', 'Gulpfile.js');
      }
      if (this.toolsExtension === 'brunch') {
        this.template('_brunch-config.js', 'brunch-config.js');
      }
      if (this.gruntcompass || this.gulpRubySass || this.sassBrunch) {
        this.directory('./demo/scss', 'app/Resources/scss');
      }
      if (this.gruntLess || this.gulpLess || this.lessBrunch) {
        this.directory('./demo/less', 'app/Resources/less');
      }
      if (this.stylusBrunch) {
        this.directory('./demo/styl', 'app/Resources/styl');
      }
      if (this.gruntcoffee || this.gulpCoffee || this.coffeeScriptBrunch) {
        this.directory('./demo/coffee', 'app/Resources/js');
      }
      if (this.gruntBabel || this.gulpBabel || this.babelBrunch) {
        this.directory('./demo/es6', 'app/Resources/js');
      }
      if (this.uglifyJsBrunch || this.gulpConcat) {
        this.directory('./demo/js', 'app/Resources/js');
      }
      if (this.gruntTypescript || this.gulpTypescript || this.typescriptBrunch) {
        this.directory('./demo/ts', 'app/Resources/js');
      }
      this.fs.copy(
        this.templatePath('_gitignore'),
        this.destinationPath('.gitignore')
      );
      this.template('_bower.json', 'bower.json');
      this.template('_package.json', 'package.json');
    },

    projectfiles: function () {
      this.fs.copy(
        this.templatePath('editorconfig'),
        this.destinationPath('.editorconfig')
      );
      this.fs.copy(
        this.templatePath('jshintrc'),
        this.destinationPath('.jshintrc')
      );
    }
  },

  checkComposer: function() {
    var done = this.async();
    this.globalComposer = false;

    child_process.execFile('composer', ['-v'], function(error, stdout, stderr) {
      if (error !== null) {
        var prompts = [{
          type: 'confirm',
          name: 'checkComposer',
          message: chalk.red('WARNING: No global composer installation found. We will install it locally if you decide to continue. Continue?'),
          default: true
        }];
        this.prompt(prompts, function (answers) {
          if (answers.checkComposer) {
            // Use the secondary installation method as we cannot assume curl is installed
            child_process.exec('php -r "readfile(\'https://getcomposer.org/installer\');" | php', function(error, stdout, stderr) {
              console.log(chalk.green('Installing composer locally.'));
              console.log('See ' + chalk.yellow('http://getcomposer.org')  + ' for more details on composer.');
              console.log('');
              this.globalComposer = true;
              done();
            }.bind(this));
          } else {
            console.log(chalk.red('Composer did not installed locally!'));
            done();
          }
        }.bind(this));
      } else {
        this.globalComposer = true;
        done();
      }
    }.bind(this));
  },

  checkBower: function() {
    this.globalBower = false;

    if (this.bootStrapSass) {
      var done = this.async();

      child_process.execFile('bower', ['-v'], function(error, stdout, stderr) {
        if (error !== null) {
          var prompts = [{
            type: 'confirm',
            name: 'checkBower',
            message: chalk.red('WARNING: No global bower installation found. We will install it locally if you decide to continue. Continue ?'),
            default: true
          }];
          this.prompt(prompts, function (answers) {
            if (answers.checkBower) {
              child_process.exec('npm install -g bower', function(error, stdout, stderr) {
                console.log(chalk.green('Installing bower locally.'));
                console.log('See ' + chalk.yellow('http://bower.io/') + ' for more details on bower.');
                console.log('');
                this.globalBower = true;
                done();
              }.bind(this));
            } else {
              console.log(chalk.red('Bower did not installed locally!'));
              done();
            }
          }.bind(this));
        } else {
          this.globalBower = true;
          done();
        }
      }.bind(this));
    }
  },

  end: {

    install: function () {
      this.installDependencies({
        bower: this.globalBower,
        npm: true,
        skipInstall: false,
        callback: function () {
          console.log(chalk.bgGreen('Everything is ready!'));
          console.log('');
        }
      });
    },

    cleanComposer: function () {
      this.spawnCommand('composer', ['remove', 'symfony/assetic-bundle']);
    },

    cleanConfig: function () {
      var done = this.async();

      var confDev = yaml.safeLoad(fs.readFileSync('app/config/config_dev.yml'));
      delete confDev.assetic;
      var newConfDev = yaml.dump(confDev, {indent: 4});
      fs.writeFileSync('app/config/config_dev.yml', newConfDev);

      var conf = yaml.safeLoad(fs.readFileSync('app/config/config.yml'));
      delete conf.assetic;
      var newConf = yaml.dump(conf, {indent: 4});
      fs.writeFileSync('app/config/config.yml', newConf);

      done();
    },

    updateAppKernel: function () {
      var appKernelPath = 'app/AppKernel.php';
      var appKernelContents = this.readFileAsString(appKernelPath);

      var newAppKernelContents = appKernelContents.replace('new Symfony\\Bundle\\AsseticBundle\\AsseticBundle(),', '');
      fs.writeFileSync(appKernelPath, newAppKernelContents);
    },

    addBundleComposer: function() {
      if (this.fixturebundle) {
        this.spawnCommand('composer', ['require', 'doctrine/doctrine-fixtures-bundle']);
      }
      if (this.migrationbundle) {
        this.spawnCommand('composer', ['require', 'doctrine/doctrine-migrations-bundle']);
      }
    },

    addBootStrapSass: function() {
      if (this.bootStrapSass) {
        this.spawnCommand('bower', ['install', 'bootstrap-sass-official', '--save']);
      }
    }
  }
});
